nexus/db-queries/src/db/datastore/rack.rs

```diff
@@ -327,8 +329,45 @@ impl DataStore {
         };
         for allocation in allocations {
             if allocation.hw_baseboard_id == new_allocation.hw_baseboard_id {
-                // We already have an allocation for this sled.
-                return Ok(SledUnderlayAllocationResult::Existing(allocation));
+                // We already have an allocation for this sled, but we need to
+                // check whether this allocation matches a sled that has been
+                // decommissioned. (The same physical sled, tracked by
+                // `hw_baseboard_id`, can be logically removed from the control
+                // plane via decommissioning, then added back again later, which
+                // requires allocating a new subnet.)
+                match LookupPath::new(opctx, self)
+                    .sled_id(allocation.sled_id.into_untyped_uuid())
+                    .optional_fetch_for(authz::Action::Read)
+                    .await?
+                    .map(|(_, sled)| sled.state())
```

Comment:
Part of me wonders if this filtering is something we should be doing on the query for subnet allocations in the first place (e.g., join with the sled table if the hw_baseboard_id matches, or else return the highest `subnet_octet + 1`), but I think it's fine to keep this as-is.

At first glance I was worried about doing a "query-per-allocation" in a loop, but I guess we're only doing that when the baseboard_id matches, which is pretty infrequent.

Reply:
It's a good question. We can't join against `hw_baseboard_id` directly, because `sled` doesn't have one (although I added a `TODO` in this PR that maybe it should?). We could left join on the `sled_id`, though? I don't have strong feels about this since as you point out, we only do this when the baseboard matches, and we only land in this function _at all_ when someone is trying to add a sled, which is itself quite infrequent.